patches/buildkit-direct-execve-v7.1/0007-fix-execvp-PATH-handling.patch

```diff
@@ -0,0 +1,73 @@
+From 511efc9c787b9d6ea00cc2c1f6a6ec781ca7bb0c Mon Sep 17 00:00:00 2001
+From: David Ackroyd <dackroyd@nine.com.au>
+Date: Tue, 22 Nov 2022 11:18:58 +1100
+Subject: [PATCH] fix execvp PATH handling
+MIME-Version: 1.0
+Content-Type: text/plain; charset=UTF-8
+Content-Transfer-Encoding: 8bit
+
+When the execvp syscall is invoked, the system PATH should be searched
+ for the executable to invoke, with the first match in the PATH being
+ invoked. However, the call being modified to inject qemu breaks this
+ behaviour, as it's not till _after_ qemu is invoked that the presence or
+ executability is checked, which is too late. ENOENT and EACCESS aren't
+ returned from the execve call, which stops execvp looping over the PATH
+ and aborts the entire process.
+
+This is resolved by testing the target command prior to executing it via
+ qemu, returning the appropriate error codes.
+
+In the case of relative paths, the prepending of the workdir resulted in
+ a corrupted string being returned such as `��@/work/./script.sh`. This
+ is because with malloc the memory isn't zeroed, so the memory is
+ 'dirty'. As a result, the strncat doesn't insert the workdir at the
+ start as desired, and an invalid path is returned.
+
+Signed-off-by: David Ackroyd <dackroyd@nine.com.au>
+---
+ linux-user/syscall.c | 15 +++++++++++++++
+ util/path.c          |  2 ++
+ 2 files changed, 17 insertions(+)
+
+diff --git a/linux-user/syscall.c b/linux-user/syscall.c
+index b3d913c851..ddac7b0e17 100644
+--- a/linux-user/syscall.c
++++ b/linux-user/syscall.c
+@@ -8840,6 +8840,21 @@ static abi_long do_syscall1(CPUArchState *cpu_env, int num, abi_long arg1,
+                 goto execve_end;
+             }
+ 
++            /*
++             * Check whether executable up front, as running once the qemu process is started these failures
++             * will happen internally there, and only exposed as a non-zero exit code for qemu.
++             */
++            ret = get_errno(stat(path(argp[3]), &st));
++            if (is_error(ret)) {
++                ret = -host_to_target_errno(errno);
++                goto execve_end;
++            }
++
++            if ((st.st_mode & S_IFDIR) || !(st.st_mode & (S_IXUSR|S_IXGRP|S_IXOTH))) {
++                ret = TARGET_EACCES;
++                goto execve_end;
++            }
++
+             /* copy guest argv1 onwards to host argv4 onwards */
+             for (gp = guest_argp + 1*sizeof(abi_ulong), q = argp + 4; gp;
+                   gp += sizeof(abi_ulong), q++) {
+diff --git a/util/path.c b/util/path.c
+index f7907b8238..14ef103aac 100644
+--- a/util/path.c
++++ b/util/path.c
+@@ -93,6 +93,8 @@ const char *prepend_workdir_if_relative(const char *path)
+     }
+     if (!(p = malloc(k * sizeof(char*)))) return NULL;
+ 
++    memset(p, '\0', sizeof(p));
```

Comment:
I wonder if we could move this into `0003-linux-user-path-in-execve-should-be-relative-to-work.patch` was just an oversight.

Reply:
Also `p[k-1] = '\0';` should be enough.

Reply:
Thanks @tiborvass, thought adding to 0003 patch would be preferable, but kept the changes contained in the 1 for the initial push. Will move it across now, with the update you've suggested.

Reply:
I'm still seeing dirty paths with `p[k-1] = '\0';`. This doesn't impact all tests though, some appear to get 'clean' (enough?) memory via the malloc, while others do not.

Example: `path` = `./printargs`, cwd = `/work`

```c
if (!(p = malloc(k * sizeof(char*)))) return NULL; // p = "PATH=/crossarch/usr/bin:/crossarch/bin:/usr/libexec/bats-core:/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:"

p[k-1] = '\0';                                     // p = "PATH=/crossarch/u"

if (!strncat(p, buf, i)) return NULL;              // p = "PATH=/crossarch/u/work"
if (!strncat(p, "/", 1)) return NULL;              // p = "PATH=/crossarch/u/work/"
if (!strncat(p, path, j)) return NULL;             // p = "PATH=/crossarch/u/work/./printargs"
```

The actual output here should have been `/work/./printargs`, but was `PATH=/crossarch/u/work/./printargs`

Did you perhaps mean `p[0] = '\0';` ?

I think that `k-1` would mean nulling the _last_ character, prior to appending the components, which doesn't seem like the intention here? The _first_ character needs to be nulled so that any existing content is ignored, and we start by inserting the workdir from the beginning ?

Using `p[0] = '\0'` instead:

```c
if (!(p = malloc(k * sizeof(char*)))) return NULL; // p = "PATH=/crossarch/usr/bin:/crossarch/bin:/usr/libexec/bats-core:/go/bin:/usr/local/go/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:"

p[0] = '\0';                                       // p = ""

if (!strncat(p, buf, i)) return NULL;              // p = "/work"
if (!strncat(p, "/", 1)) return NULL;              // p = "/work/"
if (!strncat(p, path, j)) return NULL;             // p = "/work/./printargs"
```

Reply:
Ah yeah because of how strncat works. I guess `p[0] = '\0'` before strncat is as important as `p[k-1] = '\0'` after all the strncats. Simple alternative is to use calloc instead of malloc. But probably what we should do is just use `g_build_filename`, the exact same glib function that `path()` uses in the same file. I basically reimplemented g_build_filename with a bug.

Reply:
From what I can tell from the docs, `strncat` appears to ensure NULL is added to the end automatically, so only `p[0] = '\0';` should be required, which I've pushed up now, and appears to be working